zeth_contracts/contracts/MixerBase.sol

```diff
@@ -0,0 +1,359 @@
+    //
+    //   [Root, CommitmentS, NullifierS, h_sig, h_iS, Residual Field Element(S)]
+    //
+    // ie, below is the index mapping of the primary input elements on the
+    // protoboard:
+    //
+    //   <Merkle Root>               0
+    //   <Commitment[0]>             1
+    //   ...
+    //   <Commitment[jsOut - 1]>     jsOut
+    //   <Nullifier[0]>              jsOut + 1
+    //   ...
+    //   <Nullifier[jsIn]>           jsOut + jsIn
+    //   <h_sig>                     jsOut + jsIn + 1
+    //   <Message Auth Tag[0]>       jsOut + jsIn + 2
+    //   ...
+    //   <Message Auth Tag[jsIn]>    jsOut + 2*jsIn + 1
+    //   <Residual Field Elements>   jsOut + 2*jsIn + 2
+    //
+    // The Residual field elements are structured as follows:
+    //
+    //   255                                         128         64           0
+    //   |<empty>|<h_sig>|<nullifiers>|<msg_auth_tags>|<v_pub_in>)|<v_pub_out>|
+    //
+    // where each entry entry after public output and input holds the
+    // (curve-specific) number residual bits for the corresponding 256 bit
+    // value.
+
+    // Utility function to extract a full uint256 from a field element and the
+    // n-th set of residual bits from `residual`. This function is
+    // curve-dependent.
+    function extract_bytes32(
+        uint256 field_element, uint256 residual, uint256 residual_bits_set_idx)
+        internal pure
+        returns(bytes32);
+
+    // Implementations must provide this
+    function verify_zk_proof(
+        uint256[] memory proof,
+        uint256[num_inputs] memory inputs)
+        internal
+        returns (bool);
+
+    // This function is used to extract the public values (vpub_in, vpub_out)
+    // from the residual field element(S)
+    function assemble_public_values(uint256 residual_bits)
+        public pure
+        returns (uint256 vpub_in, uint256 vpub_out)
+    {
+        // vpub_out and vpub_in occupy the first and second public_value_bits
+        vpub_out = (residual_bits & public_value_mask) * public_unit_value_wei;
+        vpub_in = ((residual_bits >> public_value_bits) & public_value_mask)
+        * public_unit_value_wei;
+    }
+
+    // This function is used to reassemble hsig given the the primary_inputs To
+    // do so, we extract the remaining bits of hsig from the residual field
+    // element(S) and combine them with the hsig field element
+    function assemble_hsig(uint256[num_inputs] memory primary_inputs) public pure
+    returns(bytes32 hsig)
+    {
+        // The h_sig residual bits are after the jsIn authentication tags and
+        // jsIn nullifier bits.
+        return extract_bytes32(
+            primary_inputs[1 + jsIn + jsOut],
+            primary_inputs[1 + jsOut + num_hash_digests],
+            2 * jsIn);
+    }
+
+    // This function is used to reassemble the nullifiers given the nullifier
+    // index [0, jsIn[ and the primary_inputs To do so, we extract the
```

Comment:
`the primary_inputs To do so,` -> `the primary_inputs. To do so,`

Reply:
fixed. thanks.